src/org/mozilla/javascript/NativeJSON.java

```diff
@@ -365,10 +355,44 @@ private static String join(Collection<Object> objs, String delimiter) {
     }
 
     private static String jo(Scriptable value, StringifyState state) {
-        if (state.stack.search(value) != -1) {
-            throw ScriptRuntime.typeErrorById("msg.cyclic.value");
+        Object trackValue = value, unwrapped = null;
```

Comment:
What do you think about the optimization of passing a StringBuilder instead of returning as String. This should reduce expensive string concatenations in the for loop at the end of the method. (can also be done in a separate PR)

Reply:
Could potentially be another PR. I don't think it's related to this one. I think string concatenation is not actually that expensive in java 9+, so maybe it's better if we can just wait it out.

This says when classes are compiled for Java 9+ to use `+` instead of StringBuilder where possible (and even gives an example of doing it in a loop.)
https://dzone.com/articles/jdk-9jep-280-string-concatenations-will-never-be-t

Another article about it:
https://medium.com/javarevisited/java-compiler-optimization-for-string-concatenation-7f5237e5e6ed

I don't 100% understand how it works. It would maybe be interesting if someone wanted to profile the differences in the current implementation when compiled for Java 11 vs Java 8, as that may show there is no need to introduce a StringBuilder in the long term, while keeping the code easier to follow.

Reply:
Not that we don't need more to do, but we could apply the same technique to bytecode generated in Rhino, by replacing string concatenations with that same target: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/invoke/StringConcatFactory.html

Today Rhino uses ConsString internally and it trips up all kinds of people who write embedding code that expects to use String. Replacing ConsString with some clever code generation would be awesome.